File: src/data/types/combatEncounter.ts
import type { TierColor } from "./tierColor.ts";

// Combat GDD §1/§2.2/§2.3/§3. Deliberately NOT a variant of
// EncounterResult (encounter.ts) -- see that file's own note on why.
// CombatEncounter instead carries its own pending -> resolved lifecycle:
// opponentThreatTier is rolled once, immediately, at detection time (the
// only randomness that happens then, per §2.4's determinism requirement);
// status/outcome only change later, via an explicit resolveCombatChoice()
// call (Agent 20's amendment) -- never automatically.
export interface CombatEncounter {
  id: string;
  voyageId: string;
  triggerContext: "travel" | "arrival";
  opponentThreatTier: TierColor;
  status: "pending" | "resolved";
  // null while pending -- only ever set by resolveCombatChoice().
  outcome: "win" | "lose" | "flee" | null;
  // null for an arrival-triggered encounter -- there is no travel window
  // to index into (§3's own note: "null for arrival-triggered").
  windowIndex: number | null;
}

File: unity/ProfitableCore/Schema/CombatEncounter.cs
namespace Profitable.Core.Schema;

// Ports src/data/types/combatEncounter.ts. Deliberately NOT a variant of
// EncounterResult -- CombatEncounter instead carries its own pending ->
// resolved lifecycle: OpponentThreatTier is rolled once, immediately, at
// detection time; Status/Outcome only change later, via an explicit
// ResolveCombatChoice call -- never automatically.
public enum CombatTriggerContext
{
    Travel,
    Arrival,
}

public enum CombatStatus
{
    Pending,
    Resolved,
}

public enum CombatOutcome
{
    Win,
    Lose,
    Flee,
}

public class CombatEncounter
{
    public string Id { get; set; } = string.Empty;
    public string VoyageId { get; set; } = string.Empty;
    public CombatTriggerContext TriggerContext { get; set; }
    public TierColor OpponentThreatTier { get; set; }
    public CombatStatus Status { get; set; }

    // Null while pending -- only ever set by ResolveCombatChoice.
    public CombatOutcome? Outcome { get; set; }

    // Null for an arrival-triggered encounter -- there is no travel
    // window to index into.
    public int? WindowIndex { get; set; }
}
